### PR TITLE
Add get_mapcoord to gammapy.maps.Geom 

### PR DESCRIPTION
**Pull request by registerrier**
At the moment `~gammapy.maps.Geom`has only a `get_coord()`methods that yields a tuple of arrays. 
It is desirable to have a MapCoord containing all coordinates contained in a Geom object.
This PR adds a `get_mapcoord()` method to do so.

### CODE REVIEW

**Comment by cdeil** [comment]
@registerrier - Thanks!

Looking around a bit, I found three methods on geom (and none on map) that return coords:
* [WcsGeom.get_coord](http://docs.gammapy.org/dev/api/gammapy.maps.WcsGeom.html#gammapy.maps.WcsGeom.get_coord)
* [WcsGeom.coord_to_tuple](http://docs.gammapy.org/dev/api/gammapy.maps.WcsGeom.html#gammapy.maps.WcsGeom.coord_to_tuple)
* [WcsGeom.pix_to_coord](http://docs.gammapy.org/dev/api/gammapy.maps.WcsGeom.html#gammapy.maps.WcsGeom.pix_to_coord)

I'm +1 to use `MapCoord` as much as possible, because there the coords have names and units, and nice convenience methods attached. If a user really needs / wants tuples, then we can add a `to_tuple` method on `MapCoord`, no?

The alternative is to have separate `get_coord` and `get_mapcoord` and similar methods returning coords forever, which is more confusing than helpful for users, no?
@registerrier @woodmd - Thoughts?

@registerrier - probably you want to get this PR in tomorrow before you go on vacation?
Do you have time to add a simple unit test with an example and some asserts on what is filled in the return value? I think some things like e.g. coords units aren't filled correctly. That's OK, you can leave TODO comment and then we improve in the coming days.

**Comment by woodmd** [comment]
I'd be in favor of changing the return type of `get_coord` to `MapCoord` rather than adding a new method.  I considered doing this earlier but held off since it entails a lot of work to update all the examples and unit tests.  Because `MapCoord` objects support tuple-style access I think the changes needed elsewhere in the code may not actually be that extensive.

File: gammapy/maps/geom.py [hunk 1]
```diff
@@ -1063,6 +1063,17 @@ def coord_to_pix(self, coords):
         """
         pass
 
+    def get_mapcoord(self):
+        """Returns a MapCoord for this `~gammapy.maps.Geom` object.
+        """
+        coord_tuple = self.get_coord()
+        cdict = dict(lon=coord_tuple[0], lat=coord_tuple[1])
```
Comment (woodmd) [inline gammapy/maps/geom.py]: I'd suggest instantiating the `MapCoord` with an `OrderedDict` object so that the axis ordering of the map is preserved.

**Comment by cdeil** [comment]
+1 to change `get_coord` to return `MapCoords`.
I think that also means that `WcsGeom.coord_to_tuple` can be deleted, possibly adapting the two callers:
```
gammapy/maps/geom.py
1214:    def coord_to_tuple(self, coord):

gammapy/maps/hpx.py
764:        c = self.coord_to_tuple(coords)

gammapy/maps/wcs.py
478:        c = self.coord_to_tuple(coords)
```

**Comment by registerrier** [comment]
I have changed the `WcsGeom.get_coord()` and `HpxGeom.get_coord()` methods to return directly a `MapCoord`. This seems to work fine since `MapCoord` supports tuple-like access. 

There are a number of remaining failed tests mostly due to `WcsNDMap.interp_by_coord()` when the latter calls `griddata`. I'll try to figure out what is going wrong latter.

**Comment by cdeil** [comment]
I've added 4a7e17f08f3f237441e89ae88e657b5efc4a814c fixing the call to scipy griddata.

Overall I'm not sure yet if we have a sane way to work with coords. This PR introduces several `tuple(map_coords)` calls the codebase and tests, and overall coords-related code is pretty complex, no? At least to me it's not clear what our strategy is where to use MapCoord or tuple objects in Gammapy, and how to review / evolve the Gammapy codebase to get there.

Is this the goal?

- The goal is to use `MapCoord` as much as possible, internally and in return values.
- We allow tuple input, but always cast to MapCoord on the first line in methods where we allow coord tuple input (this is cheap)
- We only cast to tuple where needed, usually when calling into numpy / scipy / healpy

Even if this were agreed, I'm still not sure how to get there.
One idea could be to remove `MapCoord.__iter__` and to add a method `map_coord.to_tuple()` which does this, to be more explicit and able to find all lines where we convert MapCoord to tuple.
I guess finding all methods that take coords as input can be done via code reading or search, and we add `coord = MapCoord.from_any(coord)` lines there at the top of each one (which are pass-through for MapCoord objects, and create MapCoord if tuple or dict are passed).

@woodmd @registerrier - Please review this PR and share your thoughts.

**Comment by registerrier** [comment]
Looks fine to me. Thanks for solving the issue.

I tend to agree that we would like to use mostly `MapCoords` rather than `tuple` at least externally to interact with `maps`. I don't know how much work is needed to really make everything work with `MapCoord` internally. A lot I guess.

Also I am not sure this is something desirable. `MapCoord` would be a more efficient tool if it where able to deal with `~astropy.units` and coordinates in a more general way (i.e. independently of a specific `Map` and `Geom`).  It could provide a way to interact with `Map` and `Geom` in a more unit friendly way.

File: gammapy/maps/hpx.py [hunk 2]
```diff
@@ -1687,7 +1687,13 @@ def get_idx(self, idx=None, local=False, flat=False):
 
     def get_coord(self, idx=None, flat=False):
         pix = self.get_idx(idx=idx, flat=flat)
-        return self.pix_to_coord(pix)
+        coords = self.pix_to_coord(pix)
+        cdict = OrderedDict(lon=coords[0], lat=coords[1])
```
Comment (woodmd) [inline gammapy/maps/hpx.py]: This doesn't enforce an ordering for lon/lat.  I'd suggest changing this to:
```cdict = OrderedDict([('lon',coords[0]),('lat',coords[1])])```

File: gammapy/maps/wcs.py [hunk 3]
```diff
@@ -468,7 +469,13 @@ def get_coord(self, idx=None, flat=False):
         coords = self.pix_to_coord(pix)
         if flat:
             coords = tuple([c[np.isfinite(c)] for c in coords])
-        return coords
+
+        cdict = OrderedDict(lon=coords[0], lat=coords[1])
+        for i,axis in enumerate(self.axes):
```
Comment (woodmd) [inline gammapy/maps/wcs.py]: See comment above.

File: gammapy/maps/hpx.py (hunk 2 above)
Comment (cdeil) [reply gammapy/maps/hpx.py]: Actually, the wrong order for lon / lat causes test fails in some builds, e.g. here:
https://travis-ci.org/gammapy/gammapy/jobs/373824783#L2215

I'll fix those OrderedDict calls now.

**Comment by cdeil** [comment]
> MapCoord would be a more efficient tool if it where able to deal with ~astropy.units and coordinates in a more general way (i.e. independently of a specific Map and Geom). It could provide a way to interact with Map and Geom in a more unit friendly way.

@registerrier - I think we'll just have to improve `MapCoord` step by step and the Gammapy code base as well. If you have any specific suggestions to change or add something, please open an issue any time.

**Comment by cdeil** [comment]
Fails are unrelated (https://github.com/gammapy/gammapy/pull/1403#issuecomment-386242788). Merging this now.

@registerrier - Thanks!

Obviously this is just the start, more work is needed on map coordinates.

**Pull request merged**